contracts/distributionroles.js

```diff
@@ -0,0 +1,407 @@
+    const voteIndex = dist.votes.findIndex(x => x.role === role && x.to === to && x.from === api.sender);
+    if (api.assert(voteIndex !== -1, 'already unvoted')) {
+      dist.votes.splice(voteIndex, 1);
+
+      // remove voter if no other votes
+      const votesExist = dist.votes.find(x => x.from === api.sender);
+      if (votesExist === undefined) {
+        const voterIndex = dist.voters.findIndex(x => x.account === api.sender);
+        dist.voters.splice(voterIndex, 1);
+      }
+
+      await updateStakeWeight(dist);
+      api.emit('unvote', { from: api.sender, to });
+    }
+  }
+};
+
+actions.deposit = async (payload) => {
+  const {
+    id, symbol, quantity, isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')
+    || !api.assert(quantity && api.BigNumber(quantity).dp() <= 3 && api.BigNumber(quantity).gt(0), 'invalid quantity')) {
+    return;
+  }
+
+  const dist = await api.db.findOne('batches', { _id: id });
+  const stakeToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: dist.stakeSymbol });
+  if (api.assert(dist, 'distributionroles id not found') && api.assert(dist.active, 'distributionroles must be active to deposit')
+    && api.assert(validateIncomingToken(dist, symbol), `${symbol} is not accepted by this distributionroles`)) {
+    // deposit requested tokens to contract
+    const res = await api.executeSmartContract('tokens', 'transferToContract', { symbol, quantity, to: 'distributionroles' });
+    if (res.errors === undefined
+      && res.events && res.events.find(el => el.contract === 'tokens' && el.event === 'transferToContract' && el.data.from === api.sender && el.data.to === 'distributionroles' && el.data.quantity === quantity) !== undefined) {
+      // add pending balance on contract to distribution, if any
+      const exBalance = await api.db.findOne({
+        contract: 'tokens',
+        table: 'contractsBalances',
+        query: { account: 'distributionroles', symbol },
+      });
+      const balance = exBalance !== null ? exBalance.balance + quantity : quantity;
+
+      // update weights
+      await updateStakeWeight(dist);
+
+      // distribute deposit by role
+      for (let r = 0; r < dist.roles.length; r += 1) {
+        const rolePay = api.BigNumber(balance).multipliedBy(dist.roles[r].pct).dividedBy(100).toFixed(stakeToken.precision);
+        const cands = dist.candidates.filter(cand => cand.role === dist.roles[r].name && api.BigNumber(cand.weight).gt(dist.dustWeight));
+        cands.sort((a, b) => api.BigNumber(b.weight).minus(a.weight));
+        if (cands !== undefined && cands.length > 0) {
+          const primaryCands = cands.length > dist.roles[r].primary ? cands.slice(0, dist.roles[r].primary) : cands;
+          const backupCands = cands.length > dist.roles[r].primary ? cands.slice(dist.roles[r].primary) : [];
```

Comment:
Any limitations to this? Do we want to make this randomized instead with a limit for how many we distribute?

Reply:
That leads to another question about whether to use a threshold based trigger for payouts, just like the other contract